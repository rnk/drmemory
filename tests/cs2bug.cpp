/* **********************************************************
 * Copyright (c) 2009-2010 VMware, Inc.  All rights reserved.
 * **********************************************************/

/* Dr. Memory: the memory debugger
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; 
 * version 2.1 of the License, and no later version.

 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Library General Public License for more details.

 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <iostream>

static void
test_basic()
{
    /* uninit error */
    int *p = new int;
    if (*p != 10)
        std::cout << "hi" << std::endl;

    /* unaddr error */
    int *a = new int[3];
    /* on xp64x2cpu vm, w/ VS2005 SP1, heap assert fires: detects
     * the overflow if do a[3], so doing a[4], which is not detected.
     */
    a[4] = 12;
    delete a;
}

class hasdtr {
public:
    hasdtr() { x = new int[7]; }
    ~hasdtr() { delete[] x; }
    int *x;
    int y;
    char z;
};

class parA {
public:
    parA() { a = "parA"; };
    virtual ~parA() {}
    virtual const char *getval() const { return a; }
    const char *a;
};

class parB {
public:
    parB() { b = "parB"; }
    virtual ~parB() {}
    virtual const char *getval() const { return b; }
    virtual const char *myfunc() const { return b; }
    const char *b;
};

class childAB : public parA, public parB {
public:
    childAB() { ab = "childAB"; }
    virtual ~childAB() {}
    virtual const char *getval() const { return ab; }
    const char *ab;
};

static void
test_leaks()
{
    /* test mid-chunk std::string leak (PR 535344) */
    static std::string *str = new std::string("leak");
    std::cout << "size=" << str->size() <<
        " capacity=" << str->capacity() << std::endl;

    /* test mid-chunk pointer in leak due to new[] header (PR 484544)
     * (header is only present if class has destructor)
     */
    static hasdtr *leak = new hasdtr[4];
    leak[0].y = 0;

    /* test mid-chunk pointer in leak due to multiple inheritance by
     * casting to the 2nd of the two parent classes (PR 484544)
     */
    static parB *multi = (parB *) new childAB();
    std::cout << "getval: " << multi->getval() << std::endl;
    std::cout << "myfunc: " << multi->myfunc() << std::endl;

    /* test PR 576032 (dependent leaks): std::string shouldn't show up */
#ifndef WINDOWS
    /* FIXME PR 587093: disabling on Windows until figure out why callstack messed up */
    std::string onstack = "leakme";
    static std::string *outer = new std::string(onstack);
    outer = NULL;
#endif
}

static void
test_exception()
{
   try {
       std::cout << "throwing exception" << std::endl;
       throw std::exception();
   } catch (std::exception&) {
       std::cout << "caught exception" << std::endl;
   }
}

int main() 
{
    test_leaks();

    test_basic();

    test_exception();

    std::cout << "bye" << std::endl;
}
