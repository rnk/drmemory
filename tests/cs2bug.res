# **********************************************************
# Copyright (c) 2011-2012 Google, Inc.  All rights reserved.
# Copyright (c) 2009-2010 VMware, Inc.  All rights reserved.
# **********************************************************
#
# Dr. Memory: the memory debugger
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; 
# version 2.1 of the License, and no later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Library General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
: UNINITIALIZED READ
cs2bug.cpp:83
: UNADDRESSABLE ACCESS: reading 4 byte(s)
cs2bug.cpp:91
: INVALID HEAP ARGUMENT: allocated with operator new[], freed with operator delete
cs2bug.cpp:93
memory was allocated here:
cs2bug.cpp:87
##################################################
# test_mismatch_dtr()
: INVALID HEAP ARGUMENT: allocated with operator new[], freed with operator delete
cs2bug.cpp:180
memory was allocated here:
cs2bug.cpp:178
%OPTIONAL # Linux
: INVALID HEAP ARGUMENT to free()
%ENDOPTIONAL
: INVALID HEAP ARGUMENT: allocated with operator new[], freed with free
cs2bug.cpp:183
memory was allocated here:
cs2bug.cpp:181
%OPTIONAL # only when wrapping
: UNINITIALIZED READ
cs2bug.cpp:99
cs2bug.cpp:186
%ENDOPTIONAL
: INVALID HEAP ARGUMENT: allocated with malloc, freed with operator delete
cs2bug.cpp:186
memory was allocated here:
cs2bug.cpp:184
%OPTIONAL # VS2008 Win7
: UNINITIALIZED READ
cs2bug.cpp:186
%ENDOPTIONAL
: UNADDRESSABLE ACCESS
cs2bug.cpp:189
%OPTIONAL
# MinGW xp64 crashes rather than reporting final mismatch
: UNADDRESSABLE ACCESS
cs2bug.cpp:189
%ENDOPTIONAL
%OPTIONAL
: INVALID HEAP ARGUMENT: allocated with malloc, freed with operator delete[]
cs2bug.cpp:189
memory was allocated here:
cs2bug.cpp:187
%ENDOPTIONAL
%OPTIONAL # Linux
: INVALID HEAP ARGUMENT to free()
%ENDOPTIONAL
##################################################
# test_mismatch_int()
: INVALID HEAP ARGUMENT: allocated with operator new[], freed with operator delete
cs2bug.cpp:201
memory was allocated here:
cs2bug.cpp:199
: INVALID HEAP ARGUMENT: allocated with operator new[], freed with free
cs2bug.cpp:204
memory was allocated here:
cs2bug.cpp:202
: INVALID HEAP ARGUMENT: allocated with malloc, freed with operator delete
cs2bug.cpp:207
memory was allocated here:
cs2bug.cpp:205
: INVALID HEAP ARGUMENT: allocated with malloc, freed with operator delete[]
cs2bug.cpp:210
memory was allocated here:
cs2bug.cpp:208
##################################################
# leaks
%OUT_OF_ORDER
: LEAK 4 direct bytes + 0 indirect bytes
cs2bug.cpp:82
%if UNIX
# FIXME PR 587093: string code disabled for now on Windows
: LEAK 4 direct bytes + 19 indirect bytes
cs2bug.cpp:155
%endif
%OPTIONAL # Linux/VS2005
: LEAK 88 direct bytes + 168 indirect bytes
cs2bug.cpp:178
%ENDOPTIONAL
: LEAK 88 direct bytes + 196 indirect bytes
cs2bug.cpp:181
: LEAK 7 direct bytes + 0 indirect bytes
cs2bug.cpp:187
