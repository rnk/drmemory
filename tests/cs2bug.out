# **********************************************************
# Copyright (c) 2009-2010 VMware, Inc.  All rights reserved.
# **********************************************************
#
# Dr. Memory: the memory debugger
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; 
# version 2.1 of the License, and no later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Library General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
!if WINDOWS
# if this starts not matching on other platforms/compiler just remove the print!
size=4 capacity=15
!endif
!if UNIX
size=4 capacity=4
!endif
getval: childAB
myfunc: parB
hi
bye
:::Dr.Memory::: ERRORS FOUND:
:::Dr.Memory:::       1 unique,     1 total unaddressable access(es)
:::Dr.Memory:::       1 unique,     1 total uninitialized access(es)
:::Dr.Memory:::       0 unique,     0 total invalid heap argument(s)
:::Dr.Memory:::       0 unique,     0 total warning(s)
:::Dr.Memory:::       1 total,      4 byte(s) of leak(s)
:::Dr.Memory:::       0 total,      0 byte(s) of possible leak(s)
